vshard/router/init.lua

```diff
@@ -1083,14 +1089,46 @@ local function router_group_map_callrw_args(router, bucket_ids, bucket_args)
     return grouped_args
 end
 
+--
+-- Set the appropriate mode according to bucket_ids option for backward
+-- compatibility (in case of opts_mode is nil) and check the given opts_mode
+-- correctness in other cases.
+--
+local function router_set_map_callrw_mode(opts_mode, bucket_ids)
+    if opts_mode == nil then
+        local mode = bucket_ids and MAP_CALLRW_PARTIAL or MAP_CALLRW_FULL
+        return mode, nil
+    end
+    if opts_mode == MAP_CALLRW_PARTIAL and bucket_ids == nil then
+        return nil, lerror.make('Router can\'t execute map_callrw with ' ..
+                                '\'partial\' mode and nil bucket_ids')
+    end
+    if opts_mode == MAP_CALLRW_FULL and util.table_is_numeric(bucket_ids) then
+        return nil, lerror.make('Router can\'t execute map_callrw with ' ..
+                                '\'full\' mode and numeric bucket_ids')
+    end
+    return opts_mode, nil
+end
+
 --
 -- Consistent Map-Reduce. The given function is called on masters in the cluster
 -- with a guarantee that in case of success it was executed with all buckets
 -- being accessible for reads and writes.
 --
--- The selection of masters depends on bucket_ids option. When specified, the
--- Map-Reduce is performed only on masters having at least one of these buckets.
--- Otherwise it is executed on all the masters in the cluster.
+-- The selection of masters depends on 'mode' and 'bucket_ids' options. There
+-- are 4 general scenarios how map_callrw can be executed:
+--    1) <mode = 'full', bucket_ids = nil> - user function will be executed
+--       with args on all storages.
+--    2) <mode = 'full', bucket_ids = {[bid_1] = {b_arg_1}, ...}> - storages
+--       that have at least one bucket of bucket_ids will execute user function
+--       with args and additional buckets' arguments. Other storages will
+--       execute user function only with args.
+--    3) <mode = 'partial', bucket_ids = {bid_1, ...}> - user function will be
+--       executed with args on storages that have at least one bucket of
+--       bucket_ids.
+--    4) <mode = 'partial', bucket_id = {[bid_1] = {b_arg_1}, ...}> - the same
+--       as the 3rd scenario but buckets' arguments (b_arg_1, ..., b_arg_n)
+--       will be added as additional arguments to user function.
```

Comment:
This here and in the commit message might be slightly misleading. It looks like the user should expect the bucket arguments to be passed as positional arguments like `func(all_args, bucket1_arg, bucket2_arg, ...)`. But instead it is more like `func(all_args, bucket_args_map)`. I suggest to clarify this.